File: frontend/lib/api/comments.ts
import { baseUrl, getHeaders, handleResponse } from "./client";

export interface CommentAuthor {
  firebase_uid: string;
  username: string;
  name: string;
  avatar_url: string;
}

export interface Comment {
  _id: string;
  post_id: string;
  author_id: string;
  author?: CommentAuthor;
  content: {
    text: string;
  };
  parent_comment_id: string | null;
  stats: {
    likes_count: number;
  };
  is_liked: boolean;
  replies: Comment[];
  created_at: string;
  updated_at: string;
}

export async function addComment(
  postId: string,
  text: string,
  token: string
): Promise<Comment> {
  const res = await fetch(`${baseUrl}/posts/${postId}/comments`, {
    method: "POST",
    headers: getHeaders({ token, isJson: true }),
    body: JSON.stringify({ text }),
  });

  return handleResponse<Comment>(res, "Failed to add comment");
}

export async function addReply(
  commentId: string,
  text: string,
  token: string
): Promise<Comment> {
  const res = await fetch(`${baseUrl}/comments/${commentId}/reply`, {
    method: "POST",
    headers: getHeaders({ token, isJson: true }),
    body: JSON.stringify({ text }),
  });

  return handleResponse<Comment>(res, "Failed to add reply");
}

export async function getComments(
  postId: string,
  token?: string
): Promise<Comment[]> {
  const headers = token ? getHeaders({ token, isJson: true }) : getHeaders({ isJson: true });
  const res = await fetch(`${baseUrl}/posts/${postId}/comments`, {
    method: "GET",
    headers,
  });

  return handleResponse<Comment[]>(res, "Failed to fetch comments");
}

export async function deleteComment(
  commentId: string,
  token: string
): Promise<{ success: boolean; message: string }> {
  const res = await fetch(`${baseUrl}/comments/${commentId}`, {
    method: "DELETE",
    headers: getHeaders({ token, isJson: true }),
  });

  return handleResponse<{ success: boolean; message: string }>(res, "Failed to delete comment");
}

export async function toggleCommentLike(
  commentId: string,
  token: string
): Promise<{ success: boolean; liked: boolean; likes_count: number }> {
  const res = await fetch(`${baseUrl}/comments/${commentId}/like`, {
    method: "POST",
    headers: getHeaders({ token, isJson: true }),
  });

  return handleResponse<{ success: boolean; liked: boolean; likes_count: number }>(
    res,
    "Failed to toggle comment like"
  );
}

File: backend/routes/comments.py
import logging
from typing import List, Optional
from fastapi import APIRouter, HTTPException, Depends
from utils.auth import get_current_uid
from services.comment import CommentService
from models.comment import CommentCreateRequest, CommentResponse

# [ CONFIGURATION ] ────────────────────────────────────────────────────────────
logger = logging.getLogger(__name__)
router = APIRouter(tags=["comments"])


# [ COMMENT WRITES ] ────────────────────────────────────────────────────────────

@router.post("/posts/{postId}/comments", response_model=CommentResponse)
async def add_post_comment(
    postId: str,
    data: CommentCreateRequest,
    uid: str = Depends(get_current_uid)
):
    """Broadcast a new top-level comment onto a specific post thread."""
    try:
        comment = await CommentService.add_comment(postId, uid, data)
        return comment
    except Exception:
        logger.error(f"Failed to add comment to post {postId}", exc_info=True)
        raise HTTPException(status_code=500, detail="Internal server error")


@router.post("/comments/{commentId}/reply", response_model=CommentResponse)
async def reply_to_comment(
    commentId: str,
    data: CommentCreateRequest,
    uid: str = Depends(get_current_uid)
):
    """
    Branch a reply from an existing comment.
    
    - Validates that the parent comment exists.
    - Automatically inherits the base post ID for threading.
    """
    from bson import ObjectId
    from utils.database import get_comments_collection
    try:
        # 1. Verify parent existence
        comments_col = await get_comments_collection()
        parent = await comments_col.find_one({"_id": ObjectId(commentId)})
        if not parent:
            raise HTTPException(status_code=404, detail="Parent comment not found")
        
        # 2. Append reply to thread
        reply = await CommentService.add_comment(str(parent["post_id"]), uid, data, parent_comment_id=commentId)
        return reply
    except HTTPException:
        raise
    except Exception:
        logger.error(f"Failed to reply to comment {commentId}", exc_info=True)
        raise HTTPException(status_code=500, detail="Internal server error")


@router.delete("/comments/{commentId}")
async def delete_comment(
    commentId: str,
    uid: str = Depends(get_current_uid)
):
    """
    Delete a comment and cascade removal to all child replies.
    
    - Authorization check: only authors can delete.
    """
    try:
        success = await CommentService.delete_comment(commentId, uid)
        if not success:
            raise HTTPException(status_code=404, detail="Comment not found")
        return {"success": True, "message": "Comment and its replies deleted"}
    except Exception as e:
        if str(e) == "Unauthorized":
            raise HTTPException(status_code=403, detail="You can only delete your own comments")
        logger.error(f"Failed to delete comment {commentId}", exc_info=True)
        raise HTTPException(status_code=500, detail="Internal server error")


# [ COMMENT READS ] ────────────────────────────────────────────────────────────

@router.get("/posts/{postId}/comments", response_model=List[CommentResponse])
async def get_post_comments(
    postId: str,
    uid: Optional[str] = Depends(get_current_uid) # Optional for viewing
):
    """Fetch the complete, nested comment tree for a specific post."""
    try:
        return await CommentService.fetch_comments(postId, uid)
    except Exception:
        logger.error(f"Failed to fetch comments for post {postId}", exc_info=True)
        raise HTTPException(status_code=500, detail="Internal server error")


# [ SOCIAL ENGAGEMENT ] ────────────────────────────────────────────────────────

@router.post("/comments/{commentId}/like")
async def toggle_comment_like(
    commentId: str,
    uid: str = Depends(get_current_uid)
):
    """Toggle a peer's like status on a specific comment."""
    try:
        result = await CommentService.toggle_like(commentId, uid)
        return {
            "success": True,
            **result
        }
    except Exception:
        logger.error(f"Failed to toggle like for comment {commentId}", exc_info=True)
        raise HTTPException(status_code=500, detail="Internal server error")
